### ISSUE TITLE
cycler is required, is missing, yet build succeeds.

### ISSUE THREAD

**Issue by yurivict** [matplotlib/matplotlib#5744]
Log says:
```
REQUIRED DEPENDENCIES AND EXTENSIONS
...
                cycler: yes [cycler was not found. pip will attempt to
                        install it after matplotlib.]
...
```
However, it didn't 'attempt to install' and build just succeeded.

**Comment by koobs** [comment]
It (and any other run-time dependencies) should go in setup.py:install_requires, which is currently empty:

https://github.com/matplotlib/matplotlib/blob/master/setup.py#L179

**Comment by yurivict** [comment]
Yes, so there are two problems:
* setup.py doesn't require it (as koobs said)
* build logic promises to install it and fails to act on this, or wording is unclear, because we don't use pip.

**Comment by jenshnielsen** [comment]
In fact it all works as intended. It is not a build time dependency but a run time dependency. Install requires does not install when you call `python setup.py build` but when you do `python setup.py install`. That is how the setup is supposed to work. 

I have just confirmed that by installing into a virtualenv and it works as intended. 
When you call `setup.py install` the dependencies are installed by `easy_install`.

Note that it might be preferable to install from source using `pip install .` which installs the dependencies using `pip` rather than `easy_install`.

@koobs that line is ment to be empty. It gets populated below by calls to `install_requires.extend`

**Comment by yurivict** [comment]
I don't even have pip installed. Why does the build promise that pip will do something? You shouldn't assume that everybody uses pip, because I don't -)

**Comment by jenshnielsen** [comment]
Did you bother to read what I wrote above? 

If you don't use pip the dependencies are installed by `easy_install` which is shipped by `setuptools` and `setuptools` is a dependency of `matplotlib`.

**Comment by jenshnielsen** [comment]
#5746 clarifies the message printed. Either pip or easy_install is used to install dependencies

**Comment by yurivict** [comment]
Thanks.

**Comment by tacaswell** [comment]
`pip` is included by default with py2.7.9+ and py3.4+ binaries from python.org (https://docs.python.org/3/installing/) so assuming pip is getting to be a safe assumption.   

If down-stream packagers are _removing_ functionality from the standard python installation you should file a bug with them.

**Comment by yurivict** [comment]
pip is a package manager. OS with the standard package, like FreeBSD, doesn't want to promote alternative managers doing things under the hood. This breaks installation process, and is generally unsafe.

So this isn't a bug.

**Comment by tacaswell** [comment]
I agree that you should not be using `pip` or source installing anything at the system level. `pip` should be used as a user-space package manager with venvs.  If FreeBSD does not provide those tools, that _is_ a bug.  

From my point of view (as both a user and an upstream maintainer) cross-platform package managers are wonderful.  As no matter what system I land on I know how to install what I need and mpl only needs to support a small number packing related meta-data.

I don't think we have a contact for the FreeBSD packager for mpl, do you know who that is?  When we added `cycler` as a dependency we reached out many of the down-stream packages to make sure this went smoothly.  In the near future (with in a year) we are going to pick up traitlets as a runtime dependency so the down stream packagers will need to make sure it is packaged and update the meta-data.

**Comment by yurivict** [comment]
pip isn't bundled with python on FreeBSD, it can be installed as a separate package.

Well, cross-platform package managers are convenient and easy, however in the safety department YMMV. For example npm in NodeJS doesn't have package signatures at all, therefore leaving users vulnerable to MITM attacks. Additionally, it indiscriminately downloads various versions that sometimes come direct from github, and might be untested and unsafe. In FreeBSD there is the policy that all downloads have to be fingerprinted and repeatable. I am not sure if any Linux has such requirement.

There is the listed maintainer of mpl on FreeBSD, but I did the update to 1.5.0, and added the cycler dependency. I also recently created the traitlets port that was required by jupyter. Listed maintainers don't always have time to update versions quick, and I was interested to play with mpl. I believe mpl is also in the dependency set of jupyter.
